v0/destinations/ga4/transform.js

```diff
@@ -0,0 +1,387 @@
+        );
+        break;
+      case "tutorial_complete":
+        payload.name = eventNameMapping[event.toLowerCase()];
+        payload.params = constructPayload(
+          message,
+          mappingConfig[ConfigCategory.TUTORIAL_COMPLETE.name]
+        );
+        break;
+      case "unlock_achievement":
+        payload.name = eventNameMapping[event.toLowerCase()];
+        payload.params = constructPayload(
+          message,
+          mappingConfig[ConfigCategory.UNLOCK_ACHIEVEMENT.name]
+        );
+        break;
+      case "view_search_results":
+        payload.name = eventNameMapping[event.toLowerCase()];
+        payload.params = constructPayload(
+          message,
+          mappingConfig[ConfigCategory.VIEW_SEARCH_RESULTS.name]
+        );
+        payload.params.items = getDestinationItemProperties(message, true);
+        break;
+      default:
+        logger.info(`'${event}' is a custom GA4 event`);
+        break;
+    }
+  } else {
+    // custom events category
+    // Event names are case sensitive
+    if (isReservedWebCustomEventName(event)) {
+      throw new CustomError(
+        "[Google Analytics 4] track:: Reserved custom event names are not allowd",
+        400
+      );
+    }
+
+    if (isReservedWebCustomPrefixName(event)) {
+      throw new CustomError(
+        "[Google Analytics 4] track:: Reserved custom prefix names are not allowd",
+        400
+      );
+    }
+
+    payload.name = event;
+
+    // all extra parameters passed is incorporated inside params
+    let customParameters = {};
+    customParameters = extractCustomFields(
+      message,
+      customParameters,
+      ["properties"],
+      GA4_RESERVED_PARAMETER_EXCLUSION
+    );
+    if (!isEmptyObject(customParameters)) {
+      customParameters = flattenJson(customParameters);
+      payload.params = {
+        ...payload.params,
+        ...customParameters
+      };
+    }
+    removeReservedParameterPrefixNames(payload.params);
+  }
+
+  // take GA4 user properties
+  let userProperties = {};
+  userProperties = extractCustomFields(
+    message,
+    userProperties,
+    ["user_properties"],
+    GA4_RESERVED_USER_PROPERTY_EXCLUSION
+  );
+  if (!isEmptyObject(userProperties)) {
+    userProperties = flattenJson(userProperties);
+    rawPayload.user_properties = userProperties;
+  }
+
+  removeReservedUserPropertyPrefixNames(rawPayload.user_properties);
+
+  rawPayload = { ...rawPayload, events: [payload] };
+
+  // build response
+  const response = defaultRequestConfig();
+  response.method = defaultPostRequestConfig.requestMethod;
+  response.endpoint = ENDPOINT;
+  response.headers = {
+    HOST: "www.google-analytics.com",
+    "Content-Type": "application/json"
+  };
+  response.params = {
+    api_secret: Config.apiSecret,
+    measurement_id: Config.measurementId
+  };
+  response.body.JSON = rawPayload;
+  return response;
+};
+
+const process = async event => {
```

Comment:
do we need async here?

Reply:
async doesn't make sense if you are not using promise inside a function. The feature you are mentioning its there in versionedRouter js where we make async processing of events
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

Reply:
removed